### PR TITLE
Crypto pay bot

### PR DESCRIPTION
**Pull request by Vishtar**

### CODE REVIEW

File: services/servers/crypto-pay/types/TInvoice.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
+type TPayUrl = `https://t.me/${'CryptoBot' | 'CryptoTestnetBot'}?start=${TInvoice['hash']}`
```
Comment (klntsky) [inline services/servers/crypto-pay/types/TInvoice.ts:1]: Is this file copied from somewhere?

File: services/servers/tg-bots/pics/i18n/ru.ts [hunk 2]
```diff
@@ -13,26 +13,27 @@ export const ru = {
...
-    suggestionFormat: `
+    suggestionFormat: () => `
 Необходимо отправить канал в следующем формате:
 /suggest_channel @name или https://t.me/name`,
-    suggestionThanks: 'Благодарим за предложение!',
-    error: 'Произошла ошибка. Пожалуйста, попробуйте позже',
-    noNewMemes: 'С последнего запроса не появилось новых мемов',
-    nothingFound: 'Ничего не найдено',
+    suggestionThanks: () => 'Благодарим за предложение!',
+    error: () => 'Произошла ошибка. Пожалуйста, попробуйте позже',
```
Comment (klntsky) [inline services/servers/tg-bots/pics/i18n/ru.ts:25]: Why do you delay the values with `() => ` here?

File: services/servers/tg-bots/publisher/i18n/ru.ts [hunk 3]
```diff
@@ -1,20 +1,74 @@
 export const ru = {
```
Comment (klntsky) [inline services/servers/tg-bots/publisher/i18n/ru.ts:1]: :+1:

File: utils/mysql-queries/selectRandomOcrKeyPro.ts [hunk 4]
```diff
@@ -10,7 +10,7 @@ export const selectRandomOcrKeyPro = async (db: TDbConnection) => {
     .where(
       or(
         eq(ocrKeysPro.timeout, null),
-        sql`'${ocrKeysPro.timeout}' < DATE_ADD(NOW(), INTERVAL ${OCR_SPACE_403_DELAY} MICROSECOND)`,
+        sql`'${ocrKeysPro.timeout}' < DATE_ADD(NOW(), INTERVAL ${OCR_SPACE_403_DELAY}000 MICROSECOND)`,
```
Comment (klntsky) [inline utils/mysql-queries/selectRandomOcrKeyPro.ts:13]: Is the a SECOND keyword available?

File: services/servers/crypto-pay/types/TInvoice.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+export type TInvoice = {
+  invoice_id: number
+  hash: string
+  amount: string
+  fee_asset?: 'USDT' | 'TON' | 'BTC' | 'ETH' | 'LTC' | 'BNB' | 'TRX' | 'USDC' | 'JET' // Available if status is "paid"
```
Comment (klntsky) [inline services/servers/crypto-pay/types/TInvoice.ts:7]: use string

File: services/servers/crypto-pay/types/TInvoice.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  paid_anonymously: boolean
+  comment?: string
+  hidden_message?: string
+  payload?: string
+  paid_btn_name?: 'viewItem' | 'openChannel' | 'openBot' | 'callback'
+  paid_btn_url?: string
+} & (
+  | {
+    currency_type: 'crypto'
+    asset?: 'USDT' | 'TON' | 'BTC' | 'ETH' | 'LTC' | 'BNB' | 'TRX' | 'USDC' | 'JET'
```
Comment (klntsky) [inline services/servers/crypto-pay/types/TInvoice.ts:27]: use string

File: services/servers/crypto-pay/types/TInvoice.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  }
+  | {
+    currency_type: 'fiat'
+    fiat?:
```
Comment (klntsky) [inline services/servers/crypto-pay/types/TInvoice.ts:31]: use string

File: services/servers/crypto-pay/types/TInvoice.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    | 'AMD'
+    | 'THB'
+    | 'INR'
+    | 'BRL'
+    | 'IDR'
+    | 'AZN'
+    | 'AED'
+    | 'PLN'
+    | 'ILS'
+    paid_asset?: 'USDT' | 'TON' | 'BTC' | 'ETH' | 'LTC' | 'BNB' | 'TRX' | 'USDC' | 'JET' // Available if currency_type is "fiat" and status is "paid"
```
Comment (klntsky) [inline services/servers/crypto-pay/types/TInvoice.ts:52]: use string

File: services/servers/crypto-pay/types/TInvoice.ts [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    paid_amount?: string // Available if currency_type is "fiat" and status is "paid"
+    paid_fiat_rate?: string // Available if currency_type is "fiat" and status is "paid"
+    accepted_assets?: Array<
+    'USDT' | 'TON' | 'BTC' | 'ETH' | 'LTC' | 'BNB' | 'TRX' | 'USDC' | 'JET'
```
Comment (klntsky) [inline services/servers/crypto-pay/types/TInvoice.ts:56]: use string

**Review by klntsky — CHANGES_REQUESTED** [review]

File: services/servers/tg-bots/publisher/utils/addChannel.ts [hunk 10]
```diff
@@ -36,21 +37,21 @@ export const addChannel = async (ctx, text) => {
...
       ...logEntity,
-      error: `Adding private channel`,
+      error: `Adding a private channel`,
       channel,
     })
     return
   }
-  let isOurUserAnAdmin: boolean
-  let isOurBotAnAdmin: boolean
+  let isOurUserAnAdmin: boolean | undefined
```
Comment (klntsky) [inline services/servers/tg-bots/publisher/utils/addChannel.ts:53]: Why is `undefined` possible?

File: services/servers/tg-bots/publisher/utils/handleDistributionQueue.ts [hunk 11]
```diff
@@ -32,27 +32,36 @@ export const handleDistributionQueue = async (bot: Telegraf<TTelegrafContext>, l
...
       })
 
       payload.keywords.forEach(keyword =>
         buttons.push([
-          Key.callback(
-            `🗑️ «${keyword}» (отписаться)`,
-            `${ECallback.KEY}|${EKeywordAction.DELETE}|${keyword}`,
-          ),
+          {
+            text: `🗑️ «${keyword}» (отписаться)`,
```
Comment (klntsky) [inline services/servers/tg-bots/publisher/utils/handleDistributionQueue.ts:53]: i18n?

File: services/servers/tg-bots/publisher/utils/handleInvoiceCreation.ts [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+      JSON.stringify({
+        id,
+        user,
+      }),
+    )
+    publisherToCryptoPayCh.sendToQueue(AMQP_PUBLISHER_TO_CRYPTOPAY_CHANNEL, content, {
+      persistent: true,
+    })
+  } catch (error) {
+    await ctx.reply('Что-то пошло не так. Пожалуйста, повторите попытку позже.')
```
Comment (klntsky) [inline services/servers/tg-bots/publisher/utils/handleInvoiceCreation.ts:23]: i18n?

File: utils/mysql-queries/insertPublisherPremiumUser.ts [hunk 13]
```diff
@@ -0,0 +1,13 @@
+import { botPublisherUserPremiums } from '../../db/schema'
+import { TDbConnection } from '../types'
+
+export const insertPublisherPremiumUser = async (db: TDbConnection, id: number) => {
+  const untilTimestamp = Date.now() / 1000 + 60 * 60 * 24 * 31
```
Comment (klntsky) [inline utils/mysql-queries/insertPublisherPremiumUser.ts:5]: You should allow paying forward. Get current `untilTimestamp` from the database, and increase it. For example, a user may want to pay in advance before their premium is expired.

**Review by klntsky — CHANGES_REQUESTED** [review]

File: services/servers/crypto-pay/types/TInvoice.ts (hunk 1 above)
Comment (Vishtar) [reply services/servers/crypto-pay/types/TInvoice.ts:1]: Unfortunatelly, no.

File: utils/mysql-queries/selectRandomOcrKeyPro.ts (hunk 4 above)
Comment (Vishtar) [reply utils/mysql-queries/selectRandomOcrKeyPro.ts:13]: Yes, there is a SECOND.

File: services/servers/crypto-pay/types/TInvoice.ts (hunk 5 above)
Comment (Vishtar) [reply services/servers/crypto-pay/types/TInvoice.ts:7]: Replaces these and followings in the next PR.

File: services/servers/tg-bots/publisher/utils/addChannel.ts (hunk 10 above)
Comment (Vishtar) [reply services/servers/tg-bots/publisher/utils/addChannel.ts:53]: I want to add the `"strict": true` rule into ts config. So... Let variables are undefined until we define them.

File: services/servers/tg-bots/publisher/utils/handleDistributionQueue.ts (hunk 11 above)
Comment (Vishtar) [reply services/servers/tg-bots/publisher/utils/handleDistributionQueue.ts:53]: I fixed this in the next PR.

File: utils/mysql-queries/insertPublisherPremiumUser.ts (hunk 13 above)
Comment (Vishtar) [reply utils/mysql-queries/insertPublisherPremiumUser.ts:5]: Yes, this was an MVP version. I fixed this in the next PR.

**Pull request merged**